### PR TITLE
[Blazor] Manual reflection based page discovery

### PR DESCRIPTION
**Pull request by javiercn**
* Addresses https://github.com/dotnet/aspnetcore/issues/49756, https://github.com/dotnet/aspnetcore/issues/49754
* Incorporates the API review fixes.
* Addresses https://github.com/dotnet/aspnetcore/issues/49753
  * There are small changes to the public API here that need to go through API review

### CODE REVIEW

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilderExtensions.cs
Comment (halter73) [inline src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilderExtensions.cs:22]: Did we settle on `AddAssemblies`? That's my preference and what I thought we agreed on, but you were sharing your screen, and your most recent edit of the proposal #49756 showed `AddAdditionalAssemblies`. I can update https://github.com/dotnet/aspnetcore/issues/49756#issuecomment-1660930751. I just want to make sure this is intentional.

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilder.cs
Comment (halter73) [inline src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilder.cs:14]:
```suggestion
public sealed class RazorComponentsEndpointConventionBuilder : IEndpointConventionBuilder
```

This is one of the things we made plural in https://github.com/dotnet/aspnetcore/issues/49754#issuecomment-1660809201.

File: src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentEndpointConventionBuilderExtensions.cs
Comment (halter73) [inline src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentEndpointConventionBuilderExtensions.cs:20]: This is another one where I personally prefer `WebAssemblyComponentsEndpointOptions`, but we approved `WebAssemblyComponentsEndpointsOptions` in [the API approval](https://github.com/dotnet/aspnetcore/issues/49754#issuecomment-1660809201).

File: src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt
Comment (halter73) [inline src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt:3]: This is another one where I personally prefer `WebAssemblyComponentsEndpointOptions`, but we approved `WebAssemblyComponentsEndpointsOptions` in [the API approval](https://github.com/dotnet/aspnetcore/issues/49754#issuecomment-1660809201). I think we can leave an API review note about going back to `Endpoint` if @SteveSandersonMS is fine with it.

I do think we should change the namespace though.

```suggestion
Microsoft.AspNetCore.Components.WebAssembly.Server.WebAssemblyComponentsEndpointOptions
```

File: src/Components/Endpoints/src/DependencyInjection/RazorComponentOptions.cs
Comment (halter73) [inline src/Components/Endpoints/src/DependencyInjection/RazorComponentOptions.cs:11]:
```suggestion
public sealed class RazorComponentsOptions
```

File: src/Components/Endpoints/src/PublicAPI.Unshipped.txt
Comment (halter73) [inline src/Components/Endpoints/src/PublicAPI.Unshipped.txt:17]: Can we make the `IRazorComponentsBuilder` namespace change in this PR since that's the only thing that looks to be remaining from #49754?

```suggestion
Microsoft.Extensions.DependencyInjection.IRazorComponentsBuilder
```

**Comment by halter73** [comment]
Do we expect the `RenderModeEndpointProvider` namespace change we approved in #49756 in another PR?

File: src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt
Comment (SteveSandersonMS) [reply src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt:3]: I'll go with whatever the two of you prefer.

Comment (MackinnonBuck) [reply src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt:3]: I also like `WebAssemblyComponentsEndpointOptions` (no "s"). Should we have a universal rules across these APIs that "Components" is always plural and "Endpoint" is always singular?

We might also have to change some other APIs if we decide that, though (e.g., RazorComponentsEndpoint~~s~~Options).

File: src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentEndpointConventionBuilderExtensions.cs
Comment (MackinnonBuck) [reply src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentEndpointConventionBuilderExtensions.cs]: Agreed - I've just changed it to `WebAssemblyComponentsEndpointOptions`

**Comment by MackinnonBuck** [comment]
I've just addressed some of the PR feedback about API changes. This is ready to be reviewed again.

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilderExtensions.cs
Comment (halter73) [reply src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilderExtensions.cs]: I don't understand that implication at all. We don't call `AddSingleton` `AddAdditionalSingleton` just because there are some default singleton services like `IServiceProvider`.

That said, I said basically this during API review, and I think the consensus was that we should add `Additional` to the method name to make it super clear that the main assembly is included by default.

File: src/Components/Endpoints/src/Builder/RazorComponentsEndpointConventionBuilder.cs
Comment (halter73) [inline src/Components/Endpoints/src/Builder/RazorComponentsEndpointConventionBuilder.cs:80]:
```suggestion
    public RazorComponentsEndpointConventionBuilder AddRenderMode(IComponentRenderMode renderMode)
```

Wasn't there supposed to be `AddWebAssemblyRenderMode()` directly on the `RazorComponentsEndpointConventionBuilder` according to the approved API proposal at https://github.com/dotnet/aspnetcore/issues/49754?

I don't think we approved any `IComponentRenderMode` APIs other than `RenderModeEndpointProvider` which we tried to hide in an Infrastructure namespace. I don't think `RenderMode` or the `IComponentRenderMode` interface itself was ever proposed as API either.

I see that this is being called by an `AddWebAssemblyRenderMode` extension method in `WebAssemblyRazorComponentsEndpointConventionBuilderExtensions`. I'm assuming this was done for layering reasons so we could flow options via the call to `AddWebAssemblyRenderMode(WebAssemblyComponentsEndpointOptions)`.

It's really odd to use different patterns for adding server and webassembly render modes. If we're going to do an extension method for one, we should do it for both. It's also a bit unusual to take an `WebAssemblyComponentsEndpointOptions` directly instead of an `Action<WebAssemblyComponentsEndpointOptions>`, but maybe it's okay since we're not really doing the options pattern for endpoint config. Regardless, it's something that should be discussed in API review.

File: src/Components/WebAssembly/Server/src/Builder/WebAssemblyRenderModeWithOptions.cs
Comment (halter73) [inline src/Components/WebAssembly/Server/src/Builder/WebAssemblyRenderModeWithOptions.cs:12]:
```suggestion
{
    public WebAssemblyComponentsEndpointOptions? EndpointOptions => options;
}
```

File: src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs
Comment (halter73) [inline src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:34]: Was this necessary? If so, why?

File: src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs
Comment (halter73) [inline src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:65]: `AddWebAssemblyComponents` is the much more natural place to configure options since it follows the options pattern and aligns nicely with `AddServerComponents`.

If we're going to do this last-minute non-API-approved change, shouldn't we at least test what happens if you need per-`MapRazorComponents` `WebAssmblyComponentsEndpointOptions.PathPrefix` configuration? And verify everything works if you call `MapRazorComponents` multiple times with different path prefixes. This is the kind of thing that often doesn't work if you don't test it. I had to do extra work to make `MapIdentityApi` work if you called it with multiple different `TUser` types that I would have never found without testing.

If this isn't something that is so difficult and unusual that we cannot even test it, we're better off going with the global option we had before. If we need per-`MapRazorComponents` of this in the future, we could always add a global override that works exactly like this.

**Review by halter73 — APPROVED** [review]
I'm going to approve to unblock, but we should update the API proposal to match what we're doing, and have a discussion about the major changes to AddWebAssemblyRenderMode.

File: src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt
Comment (halter73) [reply src/Components/WebAssembly/Server/src/PublicAPI.Unshipped.txt]: I like this too. We could have called it `IEndpointsRouteBuilder` or `IEndpointsConventionBuilder` but we kept `EndpointSingular` there since it's an adjective.

File: src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs
Comment (MackinnonBuck) [reply src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:34]: I don't think so, but @javiercn can correct me if I'm wrong

File: src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs
Comment (MackinnonBuck) [reply src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:65]: I just tested this locally and it works. IMO, making this an option on the endpoint is ideal because hosting multiple WebAssembly apps is something we [already support and document](https://learn.microsoft.com/aspnet/core/blazor/host-and-deploy/multiple-hosted-webassembly), and ideally we'd want an equivalent pattern with these new APIs.

I can see that there's still some deviation from the original API proposal with this PR, so I'm going to update the API review issue to match these changes and we can make a final decision on Monday.

Comment (halter73) [reply src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:65]: I'm glad it works. I'm fine with keeping it. We should probably make `AddServerRenderMode` an extension method as well for consistency. But we can merge this as-is for now and make any final adjustments after we API review it.

Comment (halter73) [reply src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:65]: And we add an automated test for this too then?

Comment (MackinnonBuck) [reply src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:65]: > And can we add an automated test for this too then?

For sure. Is it fine if we do that in a follow-up? It's actually a bit involved, and I probably won't be able to finish by EOD.

Comment (halter73) [reply src/Components/test/testassets/Components.TestServer/RazorComponentEndpointsStartup.cs:64]: That's fine with me.

File: src/Components/Endpoints/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs
Comment (eerhardt) [inline src/Components/Endpoints/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs:32]: Why was this URL changed? I believe the intention in https://github.com/dotnet/aspnetcore/pull/49377 was for this URL to be https://aka.ms/aspnet/trimming.

**Pull request merged**
